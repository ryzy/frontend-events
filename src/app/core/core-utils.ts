/**
 * Prevent re-import of core modules
 *
 * @url https://angular.io/guide/styleguide#style-04-12
 *
 * Usage:
 * ```
 * export class MyModule {
 *   constructor(@Optional() @SkipSelf() parentModule: MyModule) {
 *     throwIfAlreadyLoaded(parentModule, 'MyModule');
 *   }
 * }
 * ```
 */
export function throwIfAlreadyLoaded(parentModule: any, moduleName: string): void {
  if (parentModule) {
    throw new Error(`${moduleName} has already been loaded. Import Core modules in the AppModule only.`);
  }
}

/**
 * Returns true if code is executed in test (i.e. Karma) context.
 */
export function isUnitTestContext(): boolean {
  return !!(window as any).__karma__;
}

// tslint:disable:no-bitwise
/**
 * Generate UUID string
 */
export function uuid(length?: number): string {
  const str = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c: string) => {
    const r: number = (Math.random() * 16) | 0;
    const v: number = c === 'x' ? r : (r & 0x3) | 0x8;
    return v.toString(16);
  });

  return length ? str.substr(0, length) : str;
}
// tslint:enable:no-bitwise

/**
 * Get random number from given range
 */
export function randomRange(min: number = 0, max: number = 999): number {
  return Math.floor(Math.random() * (max - min)) + min;
}
