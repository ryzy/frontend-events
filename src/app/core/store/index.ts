import { Action, ActionReducer, ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';

import { appReducer, AppState } from './app/app-reducer';
import { AppRouterState } from './router/router';

/**
 * Store shape, for the root/core module
 * Feature stores are provided separately
 */
export interface AppRootState {
  app: AppState;
  router: fromRouter.RouterReducerState<AppRouterState>;
}

export const reducers: ActionReducerMap<AppRootState> = {
  app: appReducer as ActionReducer<AppState>,
  router: fromRouter.routerReducer as ActionReducer<fromRouter.RouterReducerState<AppRouterState>>,
};

/**
 * Log all Store actions to console.
 * To use, uncomment it in `metaReducers` below
 */
/* istanbul ignore next */
export function logger(reducer: ActionReducer<AppRootState>): ActionReducer<AppRootState> {
  return function(state: AppRootState, action: Action): AppRootState {
    console.log('%cACTION ' + action.type, 'color: blue; font-weight: bold', { action, stateBefore: state });
    return reducer(state, action);
  } as ActionReducer<AppRootState>;
}

/* istanbul ignore next */
export const metaReducers: MetaReducer<AppRootState>[] = [
  // logger
];
