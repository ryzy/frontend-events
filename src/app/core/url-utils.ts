import { Params } from '@angular/router';

import { EventsFilters } from '../event-base/model/events-filters';
import { AppSortInfo } from './model/entity';
import { AppRouterState } from './store/router/router';

export enum AppSectionUrls {
  Home = '/',
  EventsList = '/q',
}

/**
 * Make/clean params for router
 * We unset all empty variables, since we want nice/clean URLs
 */
export function makeParamsForRouter(params: Params, overrides: Params): Params {
  const newParams = { ...params, ...overrides };

  Object.keys(newParams).forEach((k) => {
    // make `true` shorter version of `1`
    if (true === newParams[k]) {
      newParams[k] = '1';
    }
    // remove all undefined's
    if (!newParams[k]) {
      delete newParams[k];
    }
  });

  return newParams;
}

export function makeEventsFiltersForRouter(filters: EventsFilters = {}): Params {
  return {
    where: filters.where || undefined,
    ws: filters.workshops ? '1' : undefined,
    fws: filters.freeWorkshops ? '1' : undefined,
  };
}

export function getEventsFiltersFromRouter(state: AppRouterState): EventsFilters {
  return (
    state && {
      where: state.params.where || '',
      workshops: !!state.params.ws,
      freeWorkshops: !!state.params.fws,
    }
  );
}

export const defaultSortInfo: AppSortInfo = {
  active: 'date',
  direction: 'asc',
};
export function makeSortInfoForRouter(sort?: AppSortInfo): Params {
  const params: Params = {
    s: undefined,
    sd: undefined,
  };

  if (sort && sort.active) {
    if (sort.active === defaultSortInfo.active && sort.direction === defaultSortInfo.direction) {
      // default sort info? return empty {}
      return params;
    } else if ('' === sort.direction) {
      // empty direction means: please remove this sorting, so we return immediately empty obj
      return params;
    } else if (sort.direction !== defaultSortInfo.direction) {
      // only generate direction for non-default direction
      params.sd = sort.direction;
    }

    params.s = sort.active;
  }

  return params;
}

/**
 * @see makeSortInfoForRouter
 */
export function getEventsSortInfoFromRouter(state: AppRouterState): AppSortInfo | undefined {
  return state && state.params.s
    ? <AppSortInfo>{
        active: state.params.s,
        direction: state.params.sd || defaultSortInfo.direction,
      }
    : undefined;
}
