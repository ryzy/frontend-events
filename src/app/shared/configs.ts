import { MatDialogConfig, MatSnackBarConfig } from '@angular/material';

/**
 * Default options for SnackBar
 */
export const matSnackBarConfig: MatSnackBarConfig = { duration: 6000 };
/**
 * Default options for confirmation Dialog modal
 */
export const matDialogConfig: MatDialogConfig = {};
