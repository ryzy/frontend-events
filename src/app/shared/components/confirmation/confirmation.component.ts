import { Component, OnInit, ChangeDetectionStrategy, Inject } from '@angular/core';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';

/**
 * Generic confirmation dialog
 */
@Component({
  selector: 'app-confirmation',
  templateUrl: './confirmation.component.html',
  styleUrls: ['./confirmation.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ConfirmationComponent implements OnInit {
  constructor(public dialogRef: MatDialogRef<ConfirmationComponent>, @Inject(MAT_DIALOG_DATA) public message: string) {}

  public ngOnInit(): void {}

  public onConfirm(): void {
    this.dialogRef.close(true);
  }

  public onCancel(): void {
    this.dialogRef.close(false);
  }
}
