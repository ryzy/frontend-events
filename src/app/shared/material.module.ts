import { NgModule } from '@angular/core';
import { MatButtonModule, MatIconModule } from '@angular/material';

@NgModule({
  exports: [
    MatButtonModule,
    MatIconModule,
    // MatToolbarModule,
    // MatCardModule,
    // MatExpansionModule,
    // MatInputModule,
    // MatTabsModule,
    // MatNativeDateModule,
    // MatSnackBarModule,
    // MatListModule,
  ]
})
export class MaterialModule {}
