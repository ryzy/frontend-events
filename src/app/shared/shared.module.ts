import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { MatButtonModule, MatCardModule, MatIconModule } from '@angular/material';

import { NotFoundPageComponent } from './containers/not-found-page/not-found-page.component';
import { UserAvatarComponent } from './components/user-avatar/user-avatar.component';

export const IMPORT_EXPORT_MODULES = [
  CommonModule,
  RouterModule,
  ReactiveFormsModule,
];

// Material Design modules (but only those which truly needs to be shared)
export const MAT_SHARED_MODULES = [
  MatIconModule,
  MatButtonModule,
  MatCardModule,
];

export const APP_COMPONENTS = [
  NotFoundPageComponent,
  UserAvatarComponent,
];

@NgModule({
  imports: [
    ...IMPORT_EXPORT_MODULES,
    ...MAT_SHARED_MODULES,
  ],
  exports: [
    ...IMPORT_EXPORT_MODULES,
    ...MAT_SHARED_MODULES,
    ...APP_COMPONENTS,
  ],
  declarations: [
    ...APP_COMPONENTS,
  ],
})
export class SharedModule {
}
