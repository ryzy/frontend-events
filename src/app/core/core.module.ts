import { NgModule, Optional, SkipSelf } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { ServiceWorkerModule } from '@angular/service-worker';
import { EffectsModule } from '@ngrx/effects';
import { RouterStateSerializer, StoreRouterConnectingModule } from '@ngrx/router-store';
import { StoreModule } from '@ngrx/store';
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { AngularFireModule } from 'angularfire2';
import { AngularFireAuthModule } from 'angularfire2/auth';
import { AngularFirestoreModule } from 'angularfire2/firestore';

import { environment } from '../../environments/environment';
import { throwIfAlreadyLoaded } from './core-utils';
import { reducers, metaReducers } from './store/index';
import { RouterEffects } from './store/router-effects';
import { AppRouterStateSerializer } from './store/router-state-serializer';

@NgModule({
  imports: [
    BrowserModule,
    BrowserAnimationsModule,

    // @ngrx store
    StoreModule.forRoot(reducers, { metaReducers }),
    StoreRouterConnectingModule,
    /* istanbul ignore next */
    !environment.production ? StoreDevtoolsModule.instrument({ name: 'what-conference-next.com' }) : [],
    EffectsModule.forRoot([RouterEffects]),

    // Firebase setup
    AngularFireModule.initializeApp(environment.firebase),
    AngularFireAuthModule,
    AngularFirestoreModule,

    // SW
    ServiceWorkerModule.register('/ngsw-worker.js', { enabled: environment.production }),
  ],
  providers: [
    { provide: RouterStateSerializer, useClass: AppRouterStateSerializer },
  ]
})
export class CoreModule {
  public constructor(@Optional() @SkipSelf() parentModule?: CoreModule) {
    throwIfAlreadyLoaded(parentModule, 'CoreModule');
  }
}
