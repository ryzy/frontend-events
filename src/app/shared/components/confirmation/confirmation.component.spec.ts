import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MAT_DIALOG_DATA, MatDialogRef } from '@angular/material/dialog';
import { SharedModule } from '../../shared.module';

import { ConfirmationComponent } from './confirmation.component';

describe('ConfirmationComponent', () => {
  let component: ConfirmationComponent;
  let fixture: ComponentFixture<ConfirmationComponent>;
  let confirmed: boolean | undefined;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [SharedModule],
      providers: [
        { provide: MAT_DIALOG_DATA, useValue: 'Some confirmation message' },
        { provide: MatDialogRef, useValue: { close: (res) => (confirmed = res) } },
      ],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ConfirmationComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should confirm', () => {
    component.onConfirm();
    expect(confirmed).toBe(true);
  });

  it('should cancel', () => {
    component.onCancel();
    expect(confirmed).toBe(false);
  });
});
