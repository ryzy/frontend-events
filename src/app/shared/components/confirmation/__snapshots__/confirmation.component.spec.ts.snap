// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationComponent should create 1`] = `
<app-confirmation
  dialogRef={[Function Object]}
  message={[Function String]}
>
  <h3>
    Are you absolutely sure?
  </h3><p>
    Some confirmation message
  </p><p>
    This change is irreversible. Are you sure you want to proceed?
  </p><p>
    <button
      class="mat-raised-button"
      mat-raised-button=""
    >
      <span
        class="mat-button-wrapper"
      >
        Cancel
      </span>
      <div
        class="mat-button-ripple mat-ripple"
        matripple=""
        ng-reflect-centered="false"
        ng-reflect-disabled="false"
        ng-reflect-trigger="[object HTMLButtonElement]"
      />
      <div
        class="mat-button-focus-overlay"
      />
    </button>
    <button
      class="mat-raised-button mat-accent"
      color="accent"
      mat-raised-button=""
      ng-reflect-color="accent"
    >
      <span
        class="mat-button-wrapper"
      >
        Sure, do it!
      </span>
      <div
        class="mat-button-ripple mat-ripple"
        matripple=""
        ng-reflect-centered="false"
        ng-reflect-disabled="false"
        ng-reflect-trigger="[object HTMLButtonElement]"
      />
      <div
        class="mat-button-focus-overlay"
      />
    </button>
  </p>
</app-confirmation>
`;
